### PR TITLE
ww3_prnc 360_day calendar support

### LINKED ISSUE #1175: WW3_PRNC does not handle 360 day calendars
**Describe the bug**
WW3 can be run using a 360 day (climate) calendar where each month has an equal number of days (30).
However, the WW3_PRNC program does not have this functionaliry meaning it is not possible to pre-process input files from netCDF files, even if calendar type is set to `360day` in the `mod_def` file.

**To Reproduce**
Generate mod_def with 360day calendar. Try running ww3_prnc on a input file with 30 day months for February.
The last day that will be processed for February is the 28th (or 29th on a leap year).

**Expected behavior**
Missing forcing data.

**Additional context**
We need to handle other calendar types when decoding dates in ww3_prnc, here:
https://github.com/NOAA-EMC/WW3/blob/3952826f7f0f79b622f5e325151355a05cda09b8/model/src/ww3_prnc.F90#L798-L808

### PR DESCRIPTION
**Pull request by ukmo-ccbunney**
# Pull Request Summary
Adds support to `ww3_prnc` for reading netCDF files with times using the "360_day" calendar.

## Description
Currently, `ww3_prnc` only supports reading netCDF files with times relative to "standard" or "gregorian" calendars even though WW3 itself can run in a "360_day" calendar.

This update adds 360_day calendar support in the Julian day calculations in `w3timemd` via subroutines `J2D` and `D2J`.
The subroutines now return/receive a pseudo-Julian day as an offset from `1800-01-01 00:00:00` if `CALTYPE .eq. '360_day'`

I was in two minds whether to modify the `J2D` and `D2J` Julian day routines to handle this as technically the returned value is not related to a Julian day if using a 360_day calendar. It seemed like the most sensible place to make the calculation though with minimal changes elsewhere in the code. However, if preferred I can move these calculations to their own separate routine, e.g. `D360_TO_DATE` and `DATE_TO_D360`.

`ww3_prnc` will check that the "calendar attribute in the file matches the expected calendar in CALTYPE (set via the `ww3_grid` MISC namelist entry) and fail if they are not the same.

There are no changes expected in the existing regtests due to this update.

### Issue(s) addressed
- fixes #1175 

### Commit Message
Adds 360_day calendar support to ww3_prnc

### Check list  
- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ ] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [ ] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested? **Local testing with 360 day netCDF input file and regtest matrix**
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) **No**
* Have the matrix regression tests been run (if yes, please note HPC and compiler)? **Yes; Cray XC; GNU Fortran**

**Regtest results**

Ignoring the usual known non-b4b regtests, the following tests had expected changes
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
ww3_tic1.4/./work_IC2IS2_IC2b                     (3 files differ)
ww3_tic1.4/./work_IC0IS2_1000                     (3 files differ)
ww3_tic1.4/./work_IC1IS2_1000                     (3 files differ)
ww3_tic1.4/./work_IC2IS2_IC2d                     (3 files differ)
ww3_tp2.14/./work_OASACM6                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (4 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ                     (2 files differ)
ww3_tp2.15/./work_ST4FLX5                     (4 files differ)
ww3_tp2.15/./work_ST6FLX5                     (3 files differ)
ww3_tp2.15/./work_MPI_5km                     (2 files differ)
ww3_tp2.15/./work_5km                     (2 files differ)
ww3_tp2.15/./work_PR3_UQ_MPI                     (2 files differ)
ww3_tp2.17/./work_ma                     (1 files differ)
ww3_tp2.17/./work_a                     (1 files differ)
ww3_tp2.17/./work_ma1                     (directory not found)
ww3_tp2.8/./work_PR3_UQ                     (1 files differ)
```

The changes are all due to a small rewording of the warning message about unset calendar attribute:
```
<      DEFAULTING TO "standard" CALENDAR
<      INPUT FILE MUST RESPECT STANDARD/GREGORIAN CALENDAR
---
>      IT MUST RESPECT STANDARD OR GREGORIAN CALENDAR
```

### CODE REVIEW

**Comment by ukmo-ccbunney** [comment]
@mickaelaccensi I would appreciate your opinion on the changes to `w3timemd` with respect to the comments I made in the PR description (specifically the `J2D` and `D2J` subroutines. I'm happy to modify the code if you would prefer to have the 360_day calendar calculations as separate routines.

File: model/src/w3timemd.F90 [hunk 1]
```diff
@@ -1313,6 +1316,21 @@ SUBROUTINE D2J(DAT,JULIAN,IERR)
...
+            (DAY - 1) +                    & 
+            HOUR / 24.0_8 +                &
+            MINUTE / 1440.0_8 +            &
+            SECOND / 86400.0_8
+
+      IERR = 0
+      RETURN
+    ENDIF
+
+    ! Standard/Gregorian calenday - return standrd Julian day calculation:
```
Comment (mickaelaccensi) [inline model/src/w3timemd.F90]: standard

File: model/src/w3timemd.F90 [hunk 2]
```diff
@@ -1450,31 +1475,38 @@ SUBROUTINE J2D(JULIAN,DAT,IERR)
     HOUR=MINUTE/60                               ! Integral hours from beginning of day
     MINUTE=MINUTE-HOUR*60                        ! Integral minutes from beginning of hour
 
-    !---------------------------------------------
-    JALPHA=IDINT((DBLE(IJUL-1867216)-0.25d0)/36524.25d0) ! Correction for Gregorian Calendar
-    JA=IJUL+1+JALPHA-IDINT(0.25d0*DBLE(JALPHA))
-    !---------------------------------------------
+    IF(CALTYPE .EQ. '360_day') THEN
+      ! Calcalate date parts for 360 day climate calendar
```
Comment (mickaelaccensi) [inline model/src/w3timemd.F90]: Calculate

File: model/src/ww3_prnc.F90 [hunk 3]
```diff
@@ -798,10 +801,22 @@ PROGRAM W3PRNC
...
-       (INDEX(CALENDAR, "gregorian").EQ.0)) THEN
-    WRITE(NDSE,1029)
+  ELSE IF ((INDEX(CALENDAR, "standard") .GT. 0) .OR. &
+        (INDEX(CALENDAR, "gregorian") .GT. 0)) THEN
+    CALENDAR = "standard"
+  ELSE IF (INDEX(CALENDAR, "360_day") .GT. 0) THEN
+    CALENDAR = "360_day"
+  ELSE
+    WRITE(NDSE,1029) CALENDAR
+    CALENDAR = "standard" ! Default to standard calendar. Better to maybe fail?
```
Comment (mickaelaccensi) [inline model/src/ww3_prnc.F90]: I'm also hesitating... If it fails, the error message should provide the following commandline to correct the missing calendar attribute :
ncatted -h -a calendar,time,c,c,'standard' file.nc

**Review by mickaelaccensi — APPROVED** [review]
I approve the changes, I've added a few comments

File: model/src/w3timemd.F90 (hunk 1 above)
Comment (ukmo-ccbunney) [reply model/src/w3timemd.F90]: You've got a good eye for spotting my typos! Thanks. :)

File: model/src/w3timemd.F90 (hunk 2 above)
Comment (ukmo-ccbunney) [reply model/src/w3timemd.F90]: Fixed!

File: model/src/ww3_prnc.F90 (hunk 3 above)
Comment (ukmo-ccbunney) [reply model/src/ww3_prnc.F90]: I think I'd prefer to make it a failure to avoid unexpected results, if you agree?

Comment (mickaelaccensi) [reply model/src/ww3_prnc.F90]: ok, it makes sense, could you add the nco command to help users to easily solve the issue ?

Comment (ukmo-ccbunney) [reply model/src/ww3_prnc.F90]: Actually - that line is the error for catching a calendar type that is not currently handled by WW3 (e.g. `calendar='no_leap'`), so we would not want the user to change it to a standard calendar (as the times would be wrong). In this case i think it is right to fail.

The case where the calendar is not set at all still defaults to "standard" with a warning.

Comment (MatthewMasarik-NOAA) [reply model/src/ww3_prnc.F90]: @ukmo-ccbunney @mickaelaccensi, I wanted to check if this has been resolved to both your satisfaction?

Comment (MatthewMasarik-NOAA) [reply model/src/ww3_prnc.F90]: Hi @ukmo-ccbunney and @mickaelaccensi, I wanted to double check this conversation has been resolved before moving forward.  Is that the case?

Comment (ukmo-ccbunney) [reply model/src/ww3_prnc.F90]: I am just waiting to hear from @mickaelaccensi to be sure he is happy with my reply above.

Comment (mickaelaccensi) [reply model/src/ww3_prnc.F90]: @ukmo-ccbunney I agree with your last comment.

**Comment by MatthewMasarik-NOAA** [comment]
Awesome, thanks @ukmo-ccbunney and @mickaelaccensi, I'll proceed with testing!

**Comment by MatthewMasarik-NOAA** [comment]
Good evening @ukmo-ccbunney, I wanted to let you know before it gets later that I'm processing your PR and I should be able to finish by our COB today.  I had a few more test diffs than you found, but I suspect that's just due to our machine differences, and they seem to be all of the type you identified

```
<      DEFAULTING TO "standard" CALENDAR
<      INPUT FILE MUST RESPECT STANDARD/GREGORIAN CALENDAR
---
>      IT MUST RESPECT STANDARD OR GREGORIAN CALENDAR
```

within the `ww3_prnc*.out` files.  I'm going to double check that's the case, and if so I can approve and merge today.

**Review by MatthewMasarik-NOAA — APPROVED** [review]
**Code review**
Pass

**Testing**
Pass

Known non-b4b's and new non-b4b's grouped below.  The new non-b4b's can all be attributed to text in `ww3_prnc*.out` logs as pointed out in the PR header.
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
## known non-b4b's
## ---------------
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (6 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (9 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (11 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (14 files differ)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_tp2.6/./work_ST4_ASCII                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)


## new non-b4b's -- verfied all due to ww3_prnc*.out
## -------------------------------------------------
ww3_tic1.4/./work_IC1IS2_1000                     (3 files differ)
ww3_tic1.4/./work_IC2IS2_IC2d                     (3 files differ)
ww3_tic1.4/./work_IC2IS2_IC2b                     (3 files differ)
ww3_tic1.4/./work_IC0IS2_1000                     (3 files differ)
ww3_tp2.14/./work_OASACM6                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (3 files differ)
ww3_tp2.15/./work_5km                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (3 files differ)
ww3_tp2.15/./work_ST6FLX5                     (3 files differ)
ww3_tp2.15/./work_PR3_UQ                     (1 files differ)
ww3_tp2.15/./work_MPI_5km                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_MPI                     (1 files differ)
ww3_tp2.15/./work_ST4FLX5                     (3 files differ)
ww3_tp2.17/./work_ma                     (3 files differ)
ww3_tp2.17/./work_a                     (3 files differ)
ww3_tp2.17/./work_mc1                     (3 files differ)
ww3_tp2.17/./work_mb                     (3 files differ)
ww3_tp2.17/./work_mc                     (3 files differ)
ww3_tp2.17/./work_ma1                     (3 files differ)
ww3_tp2.17/./work_c                     (3 files differ)
ww3_tp2.17/./work_b                     (3 files differ)
ww3_tp2.8/./work_PR3_UQ                     (1 files differ)
ww3_ufs1.1/./work_unstr_b                     (3 files differ)
ww3_ufs1.1/./work_c_npl                     (3 files differ)
ww3_ufs1.1/./work_unstr_a                     (3 files differ)
ww3_ufs1.1/./work_unstr_c                     (3 files differ)
ww3_ufs1.1/./work_d                     (3 files differ)
ww3_ufs1.1/./work_c_nth                     (3 files differ)
ww3_ufs1.1/./work_c                     (3 files differ)
ww3_ufs1.2/./work_a                     (3 files differ)
ww3_ufs1.2/./work_l                     (3 files differ)
ww3_ufs1.2/./work_c                     (3 files differ)
ww3_ufs1.2/./work_b                     (3 files differ)
ww3_ufs1.3/./work_a                     (5 files differ)

**********************************************************************
************************ identical cases *****************************
**********************************************************************
```

* [matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/14593722/matrixCompSummary.txt)
* [matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/14593723/matrixCompFull.txt)
* [matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/14593724/matrixDiff.txt)

Approved.

**Comment by MatthewMasarik-NOAA** [comment]
Thank you @ukmo-ccbunney, for expanding WW3's calendar support to include "360_day".

**Pull request merged**